lib/core/src/Cardano/Wallet/DB/Layer.hs

```diff
@@ -1395,194 +997,38 @@ deleteDelegationCertificates wid filters = do
-        Just coin -> do
-            txids <- fmap (txWithdrawalTxId . entityVal)
-                <$> selectList [ TxWithdrawalAmount >=. coin ] []
-            ms <- combineChunked (nub txids) (\chunk -> selectList
-              ((TxMetaTxId <-. chunk):txMetaFilter) [])
-            let sortTxId = case order of
-                    W.Ascending -> sortOn (Down . txMetaTxId)
-                    W.Descending -> sortOn txMetaTxId
-            let sortSlot = case order of
-                    W.Ascending -> sortOn txMetaSlot
-                    W.Descending -> sortOn (Down . txMetaSlot)
-            pure $ sortSlot $ sortTxId $ fmap entityVal ms
-
-    let txids = map txMetaTxId metas
-    (ins, cins, outs, couts, ws) <- selectTxs txids
-
-    let tip = W.currentTip cp
-
-    liftIO $ txHistoryFromEntity
-        ti tip metas ins cins outs couts ws
-  where
-    -- Note: there are sorted indices on these columns.
-    -- The secondary sort by TxId is to make the ordering stable
-    -- so that testing with random data always works.
-    sortOpt = case order of
-        W.Ascending -> [Asc TxMetaSlot, Desc TxMetaTxId]
-        W.Descending -> [Desc TxMetaSlot, Asc TxMetaTxId]
-
-selectTxMeta
-    :: W.WalletId
-    -> W.Hash "Tx"
-    -> SqlPersistT IO (Maybe TxMeta)
-selectTxMeta wid tid =
-    fmap entityVal <$> selectFirst
-        [ TxMetaWalletId ==. wid, TxMetaTxId ==. (TxId tid)]
-        [ Desc TxMetaSlot ]
-
--- | Delete the transaction, but only if it's not in ledger.
--- Returns non-zero if this was a success.
-deletePendingOrExpiredTx
-    :: W.WalletId
-    -> W.Hash "Tx"
-    -> SqlPersistT IO Int
-deletePendingOrExpiredTx wid tid = do
-    let filt = [ TxMetaWalletId ==. wid, TxMetaTxId ==. (TxId tid) ]
-    selectFirst ((TxMetaStatus ==. W.InLedger):filt) [] >>= \case
-        Just _ -> pure 0  -- marked in ledger - refuse to delete
-        Nothing -> fromIntegral <$> deleteWhereCount
-            ((TxMetaStatus <-. [W.Pending, W.Expired]):filt)
+    -> (DB.TxMeta -> Bool)
+    -> TxWalletsHistory
+    -> m [W.TransactionInfo]
+selectTxHistory cp ti wid minWithdrawal order whichMeta
+    (decorateWithTxOuts ->  TxHistoryF txs, wmetas) = do
+    tinfos <- mapM (uncurry $ mkTransactionInfo ti (W.currentTip cp)) $ do
+        TxMetaHistory metas <- maybeToList $ Map.lookup wid wmetas
+        meta <- toList metas
+        guard $  whichMeta meta
+        transaction <- maybeToList $ Map.lookup (txMetaTxId meta) txs
+        guard $ maybe
+            True
+            (\coin -> any (>= coin)
+                $ txWithdrawalAmount <$>  withdrawals transaction)
+            minWithdrawal
+        pure (transaction, meta)
+    let sortTx = case order of
+            W.Ascending -> sortOn
+                $ (,) <$> slotNo . txInfoMeta <*> Down . txInfoId
+            W.Descending -> sortOn
+                $ (,) <$> (Down . slotNo . txInfoMeta) <*> txInfoId
+    pure $ sortTx tinfos
```

Comment:
```suggestion
    pure $ sortTx tinfos
  where
    sortTx = case order of
        W.Ascending -> sortOn
            $ (,) <$> slotNo . txInfoMeta <*> Down . txInfoId
        W.Descending -> sortOn
            $ (,) <$> (Down . slotNo . txInfoMeta) <*> txInfoId
```

Putting the definition into a `where` clause makes it easier to reason about the scope of the names `order`, `txInfoMeta` and `txInfoId`. Specifically, this makes it more clear that `txInfoMeta` is a function and not a name introduced through `RecordWildcards` somewhere inside the `do` block.

Reply:
Interesting observation! Thanks